packages/logger/logger.pony

```diff
@@ -0,0 +1,103 @@
+"""
+# Logger package
+
+Provides basic logging facilities. On construction, takes 3 parameters:
+
+* LogLevel below which, no output will be logged
+* OutStream to log to
+* Lambda that converts an object of type A to a String for logging
+
+The API for using Logger is an attempt to abide by the Pony philosophy of first,
+be correct and secondly, aim for performance. One of the ways that logging can
+slow your system down is by having to evaluate expressions to be logged
+whether they will be logged or not (based on the level of logging). For example:
+
+logger.log(Warn, name + ": " + reason)
+
+will construct a new String regardless of whether we will end up logging the
+message or not.
+
+The Logger API uses boolean short circuiting to avoid this.
+
+logger(Warn) and logger.log(name + ": " + reason)
+
+will not evaluate the expression to be logged unless the log level Warn is at
+or above the overall log level for our logging. This is as close as we can get
+to zero cost for items that aren't going to end up being logged.
+
+## Example program
+
+The following program will output 'my warn message' and 'my error message' to
+STDOUT:
+
+```pony
+use "logger"
+
+actor Main
+  new create(env: Env) =>
+    let logger = Logger[String](
+      Warn,
+      env.out,
+      lambda(a: String): String => a end)
```

Comment:
It seems like this API makes the simple/common case of logging `String`s rather complex (you have to pass `String` as a type parameter to the constructor and you have to write out a trivial function that returns the only argument.  Further, I'm not really sold that it's the logger's responsibility to convert your objects to strings for you.

But supposing it is, one possible solution might be to separate into two `Logger` types.  For example, `Logger` and `ObjectLogger`.  The `Logger` would accept only strings, and only do the work of handling log levels and printing strings.  The `ObjectLogger` would have the API shown here, where you give a type parameter and a conversion function, and it invokes an internal `Logger` to handle log levels and print strings.